src/GraphQL.Tests/Types/InterfaceGraphTypeTests.cs

```diff
@@ -84,4 +300,212 @@ public MyObject()
                 .Resolve(ctx => ctx.GetArgument<string>("name"));
         }
     }
+
+    public class MySchema2 : Schema
+    {
+        public MySchema2()
+        {
+            Query = new QueryType();
+        }
+
+        public class QueryType : ObjectGraphType
+        {
+            public QueryType()
+            {
+                Name = "Query";
+
+                Field<PictureGraphType>("img");
+            }
+        }
+
+        public class PictureGraphType : ObjectGraphType
+        {
+            public PictureGraphType()
+            {
+                Name = "Picture";
+
+                Interface<ImageGraphType>();
+                Interface<ResourceGraphType>();
+                Interface<NodeGraphType>();
+
+                Field<NonNullGraphType<IdGraphType>>("id");
+                Field<StringGraphType>("url");
+                Field<StringGraphType>("thumbnail");
+                Field<BooleanGraphType>("favorite");
+            }
+        }
+
+        public class ImageGraphType : InterfaceGraphType
+        {
+            public ImageGraphType()
+            {
+                Name = "Image";
+
+                Interface<ResourceGraphType>();
+                Interface<NodeGraphType>();
+
+                Field<NonNullGraphType<IdGraphType>>("id");
+                Field<StringGraphType>("url");
+                Field<StringGraphType>("thumbnail");
+            }
+        }
+
+        public class ResourceGraphType : InterfaceGraphType
+        {
+            public ResourceGraphType()
+            {
+                Name = "Resource";
+
+                Interface<NodeGraphType>();
+
+                Field<NonNullGraphType<IdGraphType>>("id");
+                Field<StringGraphType>("url");
+            }
+        }
+
+        public class NodeGraphType : InterfaceGraphType
+        {
+            public NodeGraphType()
+            {
+                Name = "Node";
+
+                Field<NonNullGraphType<IdGraphType>>("id");
+            }
+        }
+    }
+
+    public static class TypeFirstSchema
+    {
+        public class Query
+        {
+            public Picture? Img => null!;
+        }
+
+        [Implements(typeof(INode))]
+        [Implements(typeof(IResource))]
+        [Implements(typeof(IImage))]
+        public class Picture : IImage
```

Comment:
There’s a few ways to attack this and I don’t know that I have the right answer. Currently we don’t enforce that for type-first that the .net types utilize inheritance. This makes things like adding arguments to an inherited method possible. (Although you could implement the original method privately or mark it with Ignore to mask it from GraphQL.) Also I’ve tried to keep attributes in the same relative location to the GraphQL schema. So while we could certainly and easily do what you suggested, someone reading the code would need to check all of the implemented interfaces of the CLR types to see if they are marked with GraphQLInterface or not. Further, since we don’t require marking with GraphQLInterface for other interface uses, this may be slightly confusing why it builds the interface but then doesn’t link any references to it. (Perhaps in the same way that it won’t now without ImplementsAtrribute…)

One question people might think is why doesn’t it just automatically generate GraphQL interfaces for all discovered interfaces on a CLR class. I don’t know about other people but this certainly wouldn’t work for me. I use inheritance within my code for database support and stuff that I don’t really want to expose to GraphQL. For example IHasDeleted marks any model that supports soft deletion - something that is oblivious to GraphQL callers. 

I don’t use GraphQL interfaces much. But I can see how your suggestion may be preferred, so you don’t need to constantly write Implements throughout your code, which is a bit redundant. We could add it as another option - people don’t have to use either one.

Reply:
I see. Yeah, it may be an enhancement to v8.1 then.